File: dcp-solutions/2024-12-14/solution.rb
```ruby
def nextGreaterElement(digits)
  stack = []
  # find first decreasing number from right to left 
  i = digits.size - 1
  while i >= 0 && digits[i] > digits[stack[-1]]
    i -= 1
  end
  
  return if i < 0

  # store indices of smaller numbers in stack
  j = digits.size
  until i >= 0
    j = i
    while stack.size > 0 && digits[i] > digits[stack[-1]]
      stack.pop
    end
    stack.push i
    
    i -= 1
  end

  # reverse digits from j to the right 
  tmp, i = digits[j], j + 1
  while i < digits.size
    digits[i], tmp = tmp, digits[i]
    i += 1
  end
  
  # reverse digits in stack from top to bottom with the new last number being j
  while stack.size > 0 && digits[stack[-1]] > digits[j]
    tmp, stack.pop = stack[-1], digits[stack.pop]
    digits[stack[-1]], tmp = tmp, digits[stack[-1]]
  end

  # return the result in descending order 
  digits
end
```

File: dcp-solutions/2024-12-14/solution.js
```javascript
function nextGreaterElement(digits) {
    let n = digits.length;
    let stack = [];
    
    for(let i=n-1; i>=0 ;i--){
        while(stack.length && stack[stack.length-1] <= digits[i]){
            stack.pop();
        }
        if(!stack.length){
            return [0].concat(digits);
        }
        let topElement = stack[stack.length - 1];
        if(topElement < digits[i]){
            break;
        }else{
            let j = n-1;
            while(j >= i){
                j--;
                if(digits[j] > topElement){
                    break;
                }
            }
            [digits[i], digits[j]] = [digits[j], digits[i]];
            stack.pop();
            break;
        }
    }

    for(let i=n-2;i>=0;i--){
        if(digits[i]<digits[i+1]){
            break;
        }
    }

    let j = n - 1;
    while(j >= i + 1){
        j--;
        [digits[i], digits[j]] = [digits[j], digits[i]];
    }
    return digits;
}
```